urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+          nextDiff.operation == Operation.EQUAL) {
+        // This is a single edit surrounded by equalities.
+        equality1 = prevDiff.text;
+        edit = thisDiff.text;
+        equality2 = nextDiff.text;
+
+        // First, shift the edit as far left as possible.
+        commonOffset = diff_commonSuffix(equality1, edit);
+        if (commonOffset != 0) {
+          commonString = edit.substring(edit.length() - commonOffset);
+          equality1 = equality1.substring(0, equality1.length() - commonOffset);
+          edit = commonString + edit.substring(0, edit.length() - commonOffset);
+          equality2 = commonString + equality2;
+        }
+
+        // Second, step character by character right, looking for the best fit.
+        bestEquality1 = equality1;
+        bestEdit = edit;
+        bestEquality2 = equality2;
+        bestScore = diff_cleanupSemanticScore(equality1, edit)
+            + diff_cleanupSemanticScore(edit, equality2);
+        while (edit.length() != 0 && equality2.length() != 0
+            && edit.charAt(0) == equality2.charAt(0)) {
+          equality1 += edit.charAt(0);
+          edit = edit.substring(1) + equality2.charAt(0);
+          equality2 = equality2.substring(1);
+          score = diff_cleanupSemanticScore(equality1, edit)
+              + diff_cleanupSemanticScore(edit, equality2);
+          // The >= encourages trailing rather than leading whitespace on edits.
+          if (score >= bestScore) {
+            bestScore = score;
+            bestEquality1 = equality1;
+            bestEdit = edit;
+            bestEquality2 = equality2;
+          }
+        }
+
+        if (!prevDiff.text.equals(bestEquality1)) {
+          // We have an improvement, save it back to the diff.
+          if (bestEquality1.length() != 0) {
+            prevDiff.text = bestEquality1;
+          } else {
+            pointer.previous(); // Walk past nextDiff.
+            pointer.previous(); // Walk past thisDiff.
+            pointer.previous(); // Walk past prevDiff.
+            pointer.remove(); // Delete prevDiff.
+            pointer.next(); // Walk past thisDiff.
+            pointer.next(); // Walk past nextDiff.
+          }
+          thisDiff.text = bestEdit;
+          if (bestEquality2.length() != 0) {
+            nextDiff.text = bestEquality2;
+          } else {
+            pointer.remove(); // Delete nextDiff.
+            nextDiff = thisDiff;
+            thisDiff = prevDiff;
+          }
+        }
+      }
+      prevDiff = thisDiff;
+      thisDiff = nextDiff;
+      nextDiff = pointer.hasNext() ? pointer.next() : null;
+    }
+  }
+
+  /**
+   * Given two strings, compute a score representing whether the internal
+   * boundary falls on logical boundaries.
+   * Scores range from 6 (best) to 0 (worst).
+   * @param one First string.
+   * @param two Second string.
+   * @return The score.
+   */
+  private int diff_cleanupSemanticScore(String one, String two) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1075-1093 into a separate method.